R support

jackkamm opened the pull request:
This is a PR to add R support to emacs-jupyter.

Continuing the discussion from https://github.com/dzop/emacs-jupyter/issues/87:
1) I don't know of any non-ESS R-modes that are actively used/developed. Also, my post originally used `setq` on `ess-font-lock-keywords` but I've since changed it to `setq-local` to ensure it's only local in the REPL -- does this address your concerns about this?
2) In terms of ugly output -- the R kernel returns `text/html` versions of lists and tables, which `emacs-jupyter` prints by default, for example, here is how a table gets printed
```
#+begin_src jupyter-R :kernel R :session R
  data.frame(column1=1:5, column2=6:10, columne3=11:15)
#+end_src

#+RESULTS:
#+begin_export html
<table>
<thead><tr><th scope=col>column1</th><th scope=col>column2</th><th scope=col>columne3</th></tr></thead>
<tbody>
	<tr><td>1 </td><td> 6</td><td>11</td></tr>
	<tr><td>2 </td><td> 7</td><td>12</td></tr>
	<tr><td>3 </td><td> 8</td><td>13</td></tr>
	<tr><td>4 </td><td> 9</td><td>14</td></tr>
	<tr><td>5 </td><td>10</td><td>15</td></tr>
</tbody>
</table>
#+end_export
```
Setting `:display plain` prints the table in a more readable form, but it would be best to output as an `org-element`.

jackkamm commented:
I've pushed an initial attempt to support tables in `jupyter-R.el`. The example before now looks better:
```
#+begin_src jupyter-R :kernel R :session R
  data.frame(column1=1:5, column2=6:10, columne3=11:15)
#+end_src

#+RESULTS:
| column1 | column2 | columne3 |
|---------+---------+----------|
|       1 |       6 |       11 |
|       2 |       7 |       12 |
|       3 |       8 |       13 |
|       4 |       9 |       14 |
|       5 |      10 |       15 |
```
I implemented it by overriding `jupyter-org-result` when `&context (jupyter-lang R)`. My solution uses libxml so I should also check that `(functionp 'libxml-parse-html-region)`, but I'm not sure how to add multiple checks into the `&context` block (this `cl-lib` stuff is new to me).

TODO: I'd also like to add support for lists and HTML widgets.

dzop commented:
> does this address your reservation?

If ESS is the only package used for R, its OK. I was thinking more of adding a check like

```emacs-lisp
(when <ESS is being used for R>
  (setq-local ...))
```

> I'm not sure how to add multiple checks into the `&context` block

The `&context` block only accepts a single expression so you can't combine the `jupyter-lang` specializer with something else, you will have to use something like the following

```emacs-lisp
&context ((and (eq (jupyter-kernel-language jupyter-current-client) 'R)
               (functionp 'libxml-parse-html-region))
          (eql t))
```

to check for `R` and `libxml-parse-html-region` in `&context`.  This is a little verbose, maybe it would be better just to use `(cl-call-next-method)` in the method body and fall back to a less specialized method, i.e. the default one, when `libxml-parse-html-region` is unavailable.

dzop commented on jupyter-R.el:
```diff
@@ -11,12 +11,17 @@
 
 (defun jupyter-org-result--R-html (data)
+  "Parse and convert DATA of type 'text/html' into a prettier form.
+If DATA is not a <table>, <ol>, or <dl>, returns nil."
```
We should also include `<ul>` here, right?

jackkamm replied:
I so far haven't seen the R kernel output `<ul>`, but I've added it in case.

dzop commented on jupyter-R.el:
```diff
@@ -11,12 +11,17 @@
-             (jupyter-org-result--R-html-table parsed-html))))))
+           (inner-element (jupyter-org-result--R-html-inner-element parsed-html))
+           (inner-element-type (car inner-element)))
+      (cond ((eq inner-element-type 'table)
```
We can also do

```emacs-lisp
(cl-case (car inner-element)
  (table
   ...)
  ((ol dl ul)
   ...))
```

its a little more compact.

jackkamm replied:
Nice! I've done this as you suggested.

dzop commented on jupyter-R.el:
```diff
@@ -36,6 +41,15 @@ This fixes the header column alignment when the table has row names."
 
+(defun jupyter-org-result--R-html-list (parsed-html)
+  "Renders list-like html outputs.
+Since these can have arbitrary nested structure, we don't try to parse it into
```
Can we not recursively parse out the list structure here? Or are you saying that there can be stuff in the lists that org syntax doesn't allow as being part of a list?

jackkamm replied:
I haven't extensively tested all the possibilities. But, for example, it's possible for it to return a list of tables, or list of lists of lists, etc. I'm not sure how well org can render these sorts of structures. It seemed simpler to just render the html with `shr.el` for now.

dzop replied:
OK, yeah I don't think org can handle all of that.

jackkamm commented:
Sorry @dzop , I did a squash & force push before I saw your code review. Hope it's not too confusing. I incorporated your suggestions in a subsequent commit.

This branch has reached a point that I'm more or less satisfied with, at least for an initial version. Here are some notes.

The R kernel has a few quirks that make it's behavior different. Most noticeably, tables and lists get returned in HTML form. While the tables are easy enough to convert to org, the lists can be fairly heterogeneous, and I'm not sure the correct way to convert them to org-elements. For example, we could have a list of lists of tables and lists. My solution for now is to just use `shr.el` to render lists. While this looks better than raw html, IMO `:display plain` looks much better. Ideally, we could prefer `:display plain` when rendering a list, and `:display html` when rendering a table, but I think this would add too much complexity, and I think it's fine to let the user specify the `:display` type as needed.

Another area where the R kernel seems to differ is with HTML widgets. I tested this out on the package [DT](https://rstudio.github.io/DT/), which creates a nice javascript table that can be sorted, filtered, etc in a browser, and is an important part of my data analysis workflow. Unlike other kernels, the R kernel doesn't return an `application/vnd.jupyter.widget-view+json` -- instead, it returns an HTML file, with an `isolated` metadata tag. My solution was to paste this html into a file, return a link to the file, and open it with `browse-url`.

A final issue is that `jupyter-eval-string` (and related functions) only partially work. They send code to the REPL fine, but no output is returned from the REPL (unlike Python, which displays the output in a nice message or pop-up buffer). I'm not 100% sure, but I think what's happening is that the R kernel doesn't send an `execute_result` message, which `jupyter-eval-string` uses to capture the output.

While still having some rough edges, I think `jupyter-R.el` is good enough that I can use it to do my work, so if it's alright with you I would favor merging it in and fixing the outstanding issues in future issues/PRs. But if you want me to spend a bit more time cleaning up before merging, that's fine with me too. Since you don't use R, I'd be willing to volunteer on any future issues/PRs involving `jupyter-R.el`.

dzop commented:
About `jupyter-eval-string`, it is probably because only a `:display-data` message is being sent back. It looks like I forgot to add a callback function for `:display-data` messages.

Can you set `jupyter--debug` to `t`, evaluate code that doesn't show any results when it should, and look to see if there are `:display-data` messages in the `*Messages*` buffer to verify.

jackkamm commented:
Yep, I see a `display-data` -- here's the full output from evaluating the list `1:10`:
```
SENDING: :execute-request (:code 1:10 :silent :json-false :store_history :json-false :user_expressions #s(hash-table size 1 test eql rehash-size 1.5 rehash-threshold 0.8125 data ()) :allow_stdin t :stop_on_error :json-false)
SENT: (:shell 5615fb34-99d5-4f0e-bf07-dc52d9986577)
MESSAGE: (:iopub :status (:execution_state busy))
MESSAGE: (:iopub :execute-input (:code 1:10 :execution_count 3))
DROPPING-REQ: 5d05aa7b-6098-4da3-b45f-f020df48802d
MESSAGE: (:iopub :display-data (:data (:text/html <ol class=list-inline>
	<li>1</li>
	<li>2</li>
	<li>3</li>
	<li>4</li>
	<li>5</li>
	<li>6</li>
	<li>7</li>
	<li>8</li>
	<li>9</li>
	<li>10</li>
</ol>
 :text/markdown 1. 1
2. 2
3. 3
4. 4
5. 5
6. 6
7. 7
8. 8
9. 9
10. 10


 :text/latex \begin{enumerate*}
\item 1
\item 2
\item 3
\item 4
\item 5
\item 6
\item 7
\item 8
\item 9
\item 10
\end{enumerate*}
 :text/plain  [1]  1  2  3  4  5  6  7  8  9 10) :metadata nil))
MESSAGE: (:shell :execute-reply (:status ok :execution_count 3 :payload [] :user_expressions nil))
MESSAGE: (:iopub :status (:execution_state idle))
jupyter: eval done
```

dzop commented on jupyter-R.el:
```diff
@@ -0,0 +1,111 @@
+
+(cl-defmethod jupyter-org-result ((_mime (eql :text/html)) _params data
+                                  &context (jupyter-lang R)
+                                  &optional _metadata)
```
You should change `_metadata` and `_params` to just `metadata` and `params` since you use the arguments in the method body. The underscore is used to indicate that an argument is **not** used inside the method and so that the byte compiler does not complain about unused variables.

jackkamm replied:
Done

dzop commented on jupyter-R.el:
```diff
@@ -0,0 +1,111 @@
+(require 'jupyter-repl)
+
+(cl-defmethod jupyter-repl-initialize-fontification (&context (jupyter-lang R))
+  (if (featurep 'ess)
```
Could you use `when` here.

jackkamm replied:
Fixed

dzop commented:
LGTM. Will merge after you get a chance to address my new comments. Also the issue with `jupyter-eval-string` should be resolved on the latest master.

jackkamm commented:
I've added your suggestions, rebased, squashed and force pushed. I checked and `jupyter-eval-string` works now with the R kernel. Thanks! Loving this package.

dzop commented:
Thanks!
